lib: add json schema validation

Jafaral opened the pull request.

Don-Ward commented on tests/lib/jsontests.icn:
```diff
@@ -243,3 +246,48 @@ procedure objects_equal(x,y)
+}"
+
+
+   v := jss_validate(good, schema)
+   write("schema test valid:", if v=="" then  "SUCCESS" else "FAILURE " )
+
+   v := jss_validate(bad, schema)
+   write("schema test invalid:", if v~=="" then  "SUCCESS" else "FAILURE " )
```
The helper function called by jss_validate has a large number of cases which are not explored by these two tests. Would it be possible to increase the coverage?  I'm not asking for 100% (or anywhere near) but this exercises only a tiny fraction of the code.

Jafaral replied:
of course!

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -221,15 +221,16 @@ procedure jtouf(s,strict,error)
 end
 
-#
-# It is insane to propose a single function as the primary API for the
```
I quite liked the self deprecating nature of this comment but, if it has to go then so be it.

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+procedure jsf_validate(jfname, sfname, strict, error)
+   local json_date, json_schema
+
+   jerror := ErrorHandler(error,strict)
```
should `jerror` be `error`?

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+   return _jsvalidation(json_data, json_schema, json_schema)
+end
+
+procedure jss_validate(data, schema, strict, error)
```
Need to document the parameters. (I'm guessing that this is "string validate" as opposed to "file validate".  Also make the same fix to `jerror` as above (if needed).

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+#</p>
+
+procedure validate_property(x, val)
+   z := ""
```
Although it's optional, I think we should declare local variables (c, cr, i)

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+            while i <= *req do {
+               c := 0
+               every prop := key(val) do{
+                  if prop ~== req[i] then c := c + 1
```
Is there a reason for not using augmented operators? (this comment applies to several places in this procedure)

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+            return z
+            }
+         maxProp := x["maxProperties"]
+         if \maxProp then write(space, "    maxProperties: ", left(maxProp, 20))
```
Where is `space` defined?

Jafaral replied:
dropped

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+         addItems := x["additionalItems"]
+         if \addItems then {
+            write(space, "   additionalItems: ")
+            dumpschema(x[additems])
```
where is `dumpschema` defined?

Jafaral replied:
dropped

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+                  i := i + 1
+                  }
+               if c ~= 0 then {
+                  return val || " is out of range, allowed values are: " || fullimage(enum)
```
`fullimage()` has not changed since its introduction to the repo in 2001.  Should we be using `ximage` instead?
This comment applies to several places in this procedure.

Jafaral replied:
old doesn't mean less better! ;). `ximage()` is good for debugging, whereas `fullimage()` is good for user output. Example:

ximage:
```
L199 := list(3)
   L199[1] := "Alice"
   L199[2] := "Rayya"
   L199[3] := "Rayya"
```
fullimage:
```
["Alice","Rayya","Rayya"]
```

Don-Ward replied:
Fair enough. In that case perhaps we should bring `fullimage` up to date: there are several things (e.g. patterns) that it doesn't deal with.

Jafaral replied:
add it to out growing TODO list! :)

Don-Ward commented on uni/lib/json.icn:
```diff
@@ -991,3 +998,233 @@ end
+#
+#</p>
+
+procedure _jsvalidation(obj, sch, schRoot)
```
It's probably too large a change ... but
This procedure only returns the first error it finds, leading to many cycles of "find first error" then "fix first error".

Would it be possible to suspend, rather than return and, in the calling procedures, generate a list of errors rather than just a single error?

Jafaral replied:
Yes, that makes sense. I think I will defer this for now though.

Don-Ward commented on uni/progs/jsvalid.icn:
```diff
@@ -0,0 +1,62 @@
+procedure main(args)
+   local opts, jfname, sfname, ext
+   # : means that an input is specified
+   opts := options(args, "hs:f:")
```
The error messages from `options()` do not include any clues about which options are available. I think it would be better to include a custom procedure that, at least, reminds the caller of the -h option to get help.

Jafaral replied:
Not sure I follow. we do have `usage()` and it is called on various errors and the user can still to `-h` too.

Don-Ward replied:
The point I'm making is that without a third argument to `options() `if, for example, the user types `jsvalid -help` they will get the standard error message `Unrecognized option: -e` with no clue as to what the valid options are. They have to "know" that it's `-h` or no parameters at all to print the help.  See the `usage() `proc in `uni/progs/idxgen.icn` for an example of what I'm proposing (although it perhaps doesn't need to be as elaborate as that; something that listed a summary of the options after the error message would be fine).
If you don't want to do that, that's fine too but (in my view) it makes it slightly harder to use.

Jafaral replied:
ah, well that is  easy, nothing but the best for you @Don-Ward  :)

Don-Ward reviewed (changes requested):
I think this needs further work.

cjeffery reviewed (commented):
I concur with Don's comments.  Perhaps the comment requesting better test coverage can be deferred. Most of the other suggestions are easy to fix.

Jafaral commented:
I addressed most comments.  I still need to fix the options. 

I did add more in-code comments, cleaned up, fixed, and expanded some of the validation cases.  I did add a few more tests for better coverage l. There are a lot more things that can be done, but this now provides good basic json validation.

Jafaral commented:
> I addressed most comments. I still need to fix the options.

I reviewed `options` not sure what is missing there.
